### PR TITLE
WebGL IDL and correctness.

### PR DESCRIPTION
**Pull request by emilio**
---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [x] There are tests for these changes

This is a change set with multiple intentions:
- Be correct.
- Unlock tests that are blocking @anholt.
- Ease the transition to typed arrays once the changes by @Ms2ger start rolling
  in, since I expect the amount of test expectations to update to be
  non-trivial.

I expect having it to run on try a few times, but... r? @nox or @KiChjang

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13305)

### CODE REVIEW

File: components/script/dom/webglrenderingcontext.rs [hunk 1]
```diff
@@ -376,17 +371,17 @@ impl WebGLRenderingContext {
         // if it is UNSIGNED_SHORT_5_6_5, UNSIGNED_SHORT_4_4_4_4,
         // or UNSIGNED_SHORT_5_5_5_1, a Uint16Array must be supplied.
         // If the types do not match, an INVALID_OPERATION error is generated.
-        let received_size = if let Some(data) = data {
-            if unsafe { array_buffer_view_data_checked::<u16>(data).is_some() } {
+        let received_size = if data.is_null() {
```
Comment (KiChjang) [inline components/script/dom/webglrenderingcontext.rs]: Woah, woah, woah, woah... slow down there man, what happens if data is an invalid pointer?

File: components/script/dom/webglrenderingcontext.rs [hunk 2]
```diff
@@ -797,86 +830,92 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
 
     #[allow(unsafe_code)]
     // https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.5
-    fn BufferData(&self, _cx: *mut JSContext, target: u32, data: Option<*mut JSObject>, usage: u32) {
+    fn BufferData(&self, _cx: *mut JSContext, target: u32, data: *mut JSObject, usage: u32) -> Fallible<()> {
+        if data.is_null() {
```
Comment (KiChjang) [inline components/script/dom/webglrenderingcontext.rs]: Same comments as above.

File: components/script/dom/webglrenderingcontext.rs [hunk 3]
```diff
@@ -797,86 +830,92 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
...
-            self.webgl_error(InvalidValue);
-        }
+        Ok(())
     }
 
     #[allow(unsafe_code)]
     // https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.5
-    fn BufferSubData(&self, _cx: *mut JSContext, target: u32, offset: i64, data: Option<*mut JSObject>) {
+    fn BufferSubData(&self, _cx: *mut JSContext, target: u32, offset: i64, data: *mut JSObject) -> Fallible<()> {
+        if data.is_null() {
```
Comment (KiChjang) [inline components/script/dom/webglrenderingcontext.rs]: Same here.

**Review by KiChjang — COMMENTED** [review]
I'm not quite comfortable of passing unwrapped raw pointers around, this makes our unsafety surface wider...

**Comment by anholt** [comment]
Reviewed 3 of 3 files at r1.
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[components/script/dom/webglrenderingcontext.rs, line 821 at r1](https://reviewable.io:443/reviews/servo/servo/13305#-KRuely6u_Mk8_5I39ow:-KRuely6u_Mk8_5I39ox:brgi93t) ([raw file](https://github.com/servo/servo/blob/d803194ce998bf435cc0c28f2effd2edb1180fc0/components/script/dom/webglrenderingcontext.rs#L821)):*
> ```Rust
>             handle_potential_webgl_error!(self, bound_buffer.buffer_data(target, &data_vec, usage));
>         } else {
>             // NB: array_buffer_view_to_vec should never fail when we have
> ```

I thought abv_to_vec was going to end up calling down into that !is_null assert?  I think you want a null check in this code (and I suspect we should be providing a block of zeroes if null).

---

*[components/script/dom/webglrenderingcontext.rs, line 857 at r1](https://reviewable.io:443/reviews/servo/servo/13305#-KRue0B0A65kuoAqIvci:-KRue0B0A65kuoAqIvcj:bxzc3wq) ([raw file](https://github.com/servo/servo/blob/d803194ce998bf435cc0c28f2effd2edb1180fc0/components/script/dom/webglrenderingcontext.rs#L857)):*
> ```Rust
>                 .unwrap()
>         } else {
>             // NB: array_buffer_view_to_vec should never fail when we have
> ```

Mention of Float32Array here seems funny, given that the IDL is for a BufferDataSource.  Would "(ArrayBuffer or ArrayBufferView)" mean that array_buffer_view_to_vec wouldn't fail?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/13305)*

File: components/script/dom/webglrenderingcontext.rs (hunk 1 above)
Comment (emilio) [inline components/script/dom/webglrenderingcontext.rs]: What do you mean? `data` is the pointer given by codegen, and can be null if you pass `null`.

In general, I'd be more comfortable if `object?` was represented in codegen by `Option<NonZero<*mut JSObject>>`, but...

**Comment by anholt** [comment]
*[components/script/dom/webglrenderingcontext.rs, line 374 at r3](https://reviewable.io:443/reviews/servo/servo/13305#-KRuPlP2DhUAUOo47PfS-r3-374:-KRufzM0yIEPJMQw2Ojb:b12d27) ([raw file](https://github.com/servo/servo/blob/df3b2f4fa741e8f687c0c6a64142bc49624a3203/components/script/dom/webglrenderingcontext.rs#L374)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary>
> What do you mean? `data` is the pointer given by codegen, and can be null if you pass `null`.
> 
> 
> 
> In general, I'd be more comfortable if `object?` was represented in codegen by `Option<NonZero<*mut JSObject>>`, but...
</details>
Does the abv code check that it's actually an abv that's passed in?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/13305)*

**Comment by emilio** [comment]
So, to summarize comments and discussions from this evening, since reviewable has messed up all the comments:

>  what happens if data is an invalid pointer?

Data is either a pointer to a valid `*mut JSObject` or null, I don't know what the concern is.

> I thought abv_to_vec was going to end up calling down into that !is_null assert? I think you want a null check in this code (and I suspect we should be providing a block of zeroes if null).

(Reviewable track of where the comment is, but I think it's bufferData). In buffer data `null` is fast-rejected with an invalid value per the [section 5.14.5](https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.5) of the WebGL spec.

> Mention of Float32Array here seems funny, given that the IDL is for a BufferDataSource. Would "(ArrayBuffer or ArrayBufferView)" mean that array_buffer_view_to_vec wouldn't fail?

Right, I think that's a copy-paste error on my side from other function, I'll make sure to correct the original commit, though that goes away in subsequent changes.

And yeah, once we have proper typed arrays support, that should never fail given the `array_buffer_view_xxx` functions check for both `ArrayBuffer` and `ArrayBufferView`, and `Float32Array` extends `ArrayBuffer`.

>  Does the abv code check that it's actually an abv that's passed in?

Yes, the only thing it doesn't check is for nullability of the argument, that's why that's handled separately. In the case it's not an ABV, it returns `None`. Additionally, in the `checked` versions, if the underlying type doesn't match, it also returns `None`.

---

A summary of the test failures so far:

 * We fail about 1/2 of the tests of `tex-image-and-sub-image-2d-with-array-buffer-view.html`, mostly because we don't support `FLIP_Y_WEBGL` and `PREMULTIPLY_ALPHA_WEBGL` yet. We were running no tests because of this PR because the test setup contains a `texImage2D` call with `null` as the last argument.

 * `texImage2DBadArgs.html` is fixed.
 * Still investigating the rest of them.

**Comment by emilio** [comment]
Review ping @nox?

**Comment by nox** [comment]
@emilio r=me,KiChjang

**Comment by emilio** [comment]
@bors-servo: r=nox,KiChjang

Rebased, though probably test expectations need to change after the other WebGL PRs that have landed in the meantime, let's see.

**Comment by emilio** [comment]
Updated the test expectations. There are a few new passing tests, and some new failures due to new parts of them executed, all due to GL errors in the renderer unrelated to these changes. I'll take a look at them later this evening if I find the time.

@bors-servo: r=nox,KiChjang

**Pull request merged**